lib/features/modeling/behavior/CompensateBoundaryEventBehavior.js

```diff
@@ -0,0 +1,232 @@
+import inherits from 'inherits-browser';
+
+import { getBusinessObject, is } from '../../../util/ModelUtil';
+
+import CommandInterceptor from 'diagram-js/lib/command/CommandInterceptor';
+import { hasEventDefinition, isEventSubProcess } from '../../../util/DiUtil';
+
+/**
+ * @typedef {import('diagram-js/lib/core/EventBus').default} EventBus
+ * @typedef {import('../lib/features/modeling/Modeling').default} Modeling
+ */
+
+/**
+ *
+ *
+ * @param {import('diagram-js/lib/core/EventBus').default} eventBus
+ * @param {import('../Modeling').default} modeling
+ * @param {import('../../rules/BpmnRules').default} bpmnRules
+ */
+export default function CompensateBoundaryEventBehavior(eventBus, modeling, bpmnRules) {
+
+  CommandInterceptor.call(this, eventBus);
+
+  this.preExecute('shape.replace', handleReplacement, true);
+  this.postExecuted('shape.replace', handleReplacementPostExecuted, true);
+  this.preExecute('connection.create', handleNewConnection, true);
+  this.postExecuted('connection.delete', handleConnectionRemoval, true);
+  this.postExecuted('connection.reconnect', handleReconnection, true);
+  this.postExecuted('element.updateProperties', handlePropertiesUpdate, true);
+
+  /**
+   * Given a connection from boundary event is removed, remove the forCompensation property.
+   */
+  function handleConnectionRemoval(context) {
+    const source = context.source,
+          target = context.target;
+
+    if (isCompensationBoundaryEvent(source) && isForCompensation(target)) {
+      removeIsForCompensationProperty(target);
+    }
+  }
+
+  /**
+   * Add forCompensation property and make sure only a single compensation activity is connected.
+   */
+  function handleNewConnection(context) {
+    const source = context.source,
+          target = context.target;
+
+    if (isCompensationBoundaryEvent(source) && canBeForCompensation(target)) {
+      addIsForCompensationProperty(target);
+      removeExistingAssociation(source);
+    }
+  }
+
+  function handleReconnection(context) {
+    const newSource = context.newSource,
+          newTarget = context.newTarget,
+          oldSource = context.oldSource,
+          oldTarget = context.oldTarget;
+
+    // source changes
+    if (oldSource !== newSource) {
+      const target = oldTarget;
+
+      // newSource no longer valid
+      if (isForCompensation(target) && !isCompensationBoundaryEvent(newSource)) {
+        removeIsForCompensationProperty(target);
+      }
+
+      // newSource valid
+      if (canBeForCompensation(target) && isCompensationBoundaryEvent(newSource)) {
+        addIsForCompensationProperty(target);
```

Comment:
What scenario did we not cover here?

Reply:
OK so I debugged it and source changes are handled via ReplaceConnectionBehavior. We don't need explicit compensation-specific code.